### PR TITLE
Add coin::value_mut function for gas optimize propose.

### PR DESCRIPTION
**Pull request by e99243506bigplay**
### Description
in code comment.
### Test Plan

Test case is provided.

I unable to write a "spec" that I don't know how to install move prover dependency.
After the PR opens, I am going report bug from line 893-894.

EDIT: I am somewhat unable to use prover on this error infomation after executed `/script/dev_setup.sh` .
```
{
  "Error": "Move Prover failed: No boogie executable set.  Please set BOOGIE_EXE"
}
```

### CODE REVIEW

**Comment by gregnazario** [comment]
@e99243506bigplay did you follow the prover instructions here? https://aptos.dev/cli-tools/install-move-prover

**Comment by e99243506bigplay** [comment]
> @e99243506bigplay did you follow the prover instructions here? https://aptos.dev/cli-tools/install-move-prover

Try to following now.

**Comment by e99243506bigplay** [comment]
> @e99243506bigplay did you follow the prover instructions here? https://aptos.dev/cli-tools/install-move-prover

Results like that:
Random Pastebin: https://defuse.ca/b/rVPZ989n715amlFTVp7AjO

**Comment by e99243506bigplay** [comment]
> > @e99243506bigplay did you follow the prover instructions here? https://aptos.dev/cli-tools/install-move-prover
> 
> Results like that: Random Pastebin: https://defuse.ca/b/rVPZ989n715amlFTVp7AjO

Problem is resolved, DO NOT install Boogie 2.16.0, instead to install Boogie 2.15.8. try to using `move prover` now.

**Comment by gregnazario** [comment]
> > > @e99243506bigplay did you follow the prover instructions here? https://aptos.dev/cli-tools/install-move-prover
> > 
> > 
> > Results like that: Random Pastebin: https://defuse.ca/b/rVPZ989n715amlFTVp7AjO
> 
> Problem is resolved, DO NOT install Boogie 2.16.0, instead to install Boogie 2.15.8. try to using `move prover` now.

Interesting... Yeah, I don't run the move prover all that often myself.  If that's the case we should put a warning about that.

**Comment by e99243506bigplay** [comment]
> > > > @e99243506bigplay did you follow the prover instructions here? https://aptos.dev/cli-tools/install-move-prover
> > > 
> > > 
> > > Results like that: Random Pastebin: https://defuse.ca/b/rVPZ989n715amlFTVp7AjO
> > 
> > 
> > Problem is resolved, DO NOT install Boogie 2.16.0, instead to install Boogie 2.15.8. try to using `move prover` now.
> 
> Interesting... Yeah, I don't run the move prover all that often myself. If that's the case we should put a warning about that.

Yea, its working now, just use the Boogie 2.15.8 then all okay. however I don't see the point to write a spec file on coin::value_mut so I am skip to write a spec on coin::value_mut

**Comment by davidiw** [comment]
What are the gas savings here? This feels more like a compiler tweak, though I know those are a bit out...

**Comment by e99243506bigplay** [comment]
> What are the gas savings here? This feels more like a compiler tweak, though I know those are a bit out...

Current move compiler unable pass &mut reference as & reference, if do so, compiler will get error.

```
module me::test2 {
    use std::coin::{Self, Coin};
    use std::signer::address_of;
    struct DemoStruct<phantom CoinType> has key {
        coin: Coin<CoinType>
    }

    native public fun value_mut<CoinType>(coin: &mut Coin<CoinType>): u64;

    fun main<A>(s: &signer): Coin<A> acquires DemoStruct {
          let coin_mut = borrow_global_mut<DemoStruct<A>>(address_of(s));
          let value_now = coin::value(&coin_mut.coin);
          let new_coin = coin::extract(&mut coin_mut.coin, value_now);
          new_coin
    }

    fun main2<A>(s: &signer): Coin<A> acquires DemoStruct {
          let coin_mut = &mut borrow_global_mut<DemoStruct<A>>(address_of(s)).coin;
          let value_now = value_mut(coin_mut);
          let new_coin = coin::extract(coin_mut, value_now);
          new_coin
    }
}
```

Result:

```
// Move bytecode v5
module d1b58e44ea11ffd326a280c453b080fd.test2 {
use 8af294298815a00e65b6e2bffc48d6ac::coin;
use 8af294298815a00e65b6e2bffc48d6ac::signer;


struct DemoStruct<phantom Ty0> has key {
	coin: Coin<Ty0>
}

main<Ty0>(Arg0: &signer): Coin<Ty0> {
L0:	loc1: u64
B0:
	0: MoveLoc[0](Arg0: &signer)
	1: Call signer::address_of(&signer): address
	2: MutBorrowGlobalGeneric[0](DemoStruct<Ty0>)
	3: StLoc[1](loc0: &mut DemoStruct<Ty0>)
	4: CopyLoc[1](loc0: &mut DemoStruct<Ty0>)
	5: ImmBorrowFieldGeneric[0](DemoStruct.coin: Coin<Ty0>)
	6: Call coin::value<Ty0>(&Coin<Ty0>): u64
	7: StLoc[2](loc1: u64)
	8: MoveLoc[1](loc0: &mut DemoStruct<Ty0>)
	9: MutBorrowFieldGeneric[0](DemoStruct.coin: Coin<Ty0>)
	10: MoveLoc[2](loc1: u64)
	11: Call coin::extract<Ty0>(&mut Coin<Ty0>, u64): Coin<Ty0>
	12: Ret
}
main2<Ty0>(Arg0: &signer): Coin<Ty0> {
L0:	loc1: u64
B0:
	0: MoveLoc[0](Arg0: &signer)
	1: Call signer::address_of(&signer): address
	2: MutBorrowGlobalGeneric[0](DemoStruct<Ty0>)
	3: MutBorrowFieldGeneric[0](DemoStruct.coin: Coin<Ty0>)
	4: StLoc[1](loc0: &mut Coin<Ty0>)
	5: CopyLoc[1](loc0: &mut Coin<Ty0>)
	6: Call value_mut<Ty0>(&mut Coin<Ty0>): u64
	7: StLoc[2](loc1: u64)
	8: MoveLoc[1](loc0: &mut Coin<Ty0>)
	9: MoveLoc[2](loc1: u64)
	10: Call coin::extract<Ty0>(&mut Coin<Ty0>, u64): Coin<Ty0>
	11: Ret
}
native public value_mut<Ty0>(Arg0: &mut Coin<Ty0>): u64
}


```

The different is exists.

File: aptos-move/framework/aptos-framework/sources/coin.move [hunk 1]
```diff
@@ -528,9 +528,9 @@ module aptos_framework::coin {
     /// This function to resolve your problem, and for the gas optimize.
     /// Before:
     /// ```
-    /// let coin_mut = borrow_global_mut<StructName<CoinType>>(my_addr).coin;
-    /// let value_now = coin::value(&coin_mut);
-    /// let new_coin = coin::extract(&mut coin_mut, value_now - 100000000);
+    /// let coin_mut = borrow_global_mut<StructName<CoinType>>(my_addr);
```
Comment (wrwg) [inline aptos-move/framework/aptos-framework/sources/coin.move:531]: If there is a bug in the compiler can we please open an issue with a minimal example to reproduce the bug? In another context, I asked @e99243506bigplay already to do this, but did not got a reply.

The compiler has no issue in treating a `&mut T` as a `&T`. The following just works fine:

```
   struct C has key {
        v: u64
    }

    fun value(c: &C): u64 {
        c.v
    }

    fun foo():u64 acquires C {
        let mut_c = borrow_global_mut<C>(@42);
        let v = value(mut_c);
        v + 1
    }
```

I guess there was a different mistake in @e99243506bigplay code which let to they could not use the `&mut`? Borrow semantics is not easy. Perhaps asking with the according example which does not compile the community on Discord is a good idea?

As regards the gas optimization I do not think we want to make those APIs more heavy and confusing by adding new functions for marginal gas optimizations. Instead, we should (and will) invest soon in better code optimizations.

Comment (e99243506bigplay) [reply aptos-move/framework/aptos-framework/sources/coin.move:531]: which reply I didnt got received? let me test again.
Strange, I was thought it wont get compiled... How could be this result...

Comment (e99243506bigplay) [reply aptos-move/framework/aptos-framework/sources/coin.move:531]: > I guess there was a different mistake in @e99243506bigplay code which let to they could not use the &mut? Borrow semantics is not easy. Perhaps asking with the according example which does not compile the community on Discord is a good idea?

I will do in next time, thanks.

**Comment by clay-aptos** [comment]
> > > > @e99243506bigplay did you follow the prover instructions here? https://aptos.dev/cli-tools/install-move-prover
> > > 
> > > 
> > > Results like that: Random Pastebin: https://defuse.ca/b/rVPZ989n715amlFTVp7AjO
> > 
> > 
> > Problem is resolved, DO NOT install Boogie 2.16.0, instead to install Boogie 2.15.8. try to using `move prover` now.
> 
> Interesting... Yeah, I don't run the move prover all that often myself. If that's the case we should put a warning about that.

@gregnazario , should we add a warning about Boogie version? @wrwg and Move team, can you confirm this is the case? Thanks!

**Pull request closed without merging**